Parser

bobheadxi opened the pull request:
:tickets: **Ticket(s)**: for #26 

---

## :construction_worker: Changes

wip parser

## :flashlight: Testing Instructions

```
cmake .
make durian_test
```

rwblickhan committed:
Add single-character helpers to lexer

rwblickhan committed:
Finish lexer

rwblickhan committed:
Stub out parser

bobheadxi committed:
Scaffold Parser, ASTree, and Statements

bobheadxi committed:
Simplify Lexer::isAlpha and Lexer::isDigit

bobheadxi committed:
WIP parser - scaffolds and rudimentary literal parsing

bobheadxi committed:
Update gitignore with CMake output

CoffeeTableEspresso commented on src/include/ASTree.h:
```diff
@@ -0,0 +1,108 @@
+//////////////////////////////////////////
+/// Functions
+
+class Function : public Statement {
```
Functions should inherit from Expression, since functions can return values.

CoffeeTableEspresso commented on src/include/Lexer.h:
```diff
@@ -5,13 +5,16 @@
+private:
+    bool isDigit(const char c);
+    bool isAlpha(const char c);
+    bool isIdentChar(const char c);
```
Looks good!

CoffeeTableEspresso commented on src/include/Parser.h:
```diff
@@ -0,0 +1,45 @@
+class Parser {
+    std::unique_ptr<ASTree> m_tree;
+    Lexer m_lexer;
+    Token m_currentToken;
```
Why do you need to store any AST stuff in the parser as a field?

bobheadxi replied:
do you mean the tree, or `m_lexer` and `m_currentToken` as well?

CoffeeTableEspresso commented on src/include/Parser.h:
```diff
@@ -0,0 +1,45 @@
+    Token m_currentToken;
+
+public:
+    Parser(const std::string input);
```
Constructor should take in a Lexer, Lexer should take in a string

CoffeeTableEspresso commented on src/include/Parser.h:
```diff
@@ -0,0 +1,45 @@
+    void buildTree();
+private:
+    Token getNextToken() {
+        return m_currentToken = m_lexer.getToken();
```
Nice!

bobheadxi committed:
Remove ASTree, inject Lexer dependency, update test

Addressing feedback.

CoffeeTableEspresso commented on src/parser/Parser.cpp:
```diff
@@ -0,0 +1,57 @@
+#include <Parser.h>
+
+Parser::Parser(const std::string input)
```
this should be `explicit`

bobheadxi committed:
Add explicit to single-argument constructor

rwblickhan commented on CMakeLists.txt:
```diff
@@ -8,10 +8,11 @@ include_directories(${PROJECT_SOURCE_DIR}/src/include)
 
 set(TEST_FILES
-        test/TestLexer.cpp)
+        test/TestLexer.cpp test/TestParser.cpp src/include/ASTree.h)
```
You don't need to include `.h` files in the CMakeLists.

bobheadxi replied:
ah, i was just blindly hitting "ok" on the CLion prompts

rwblickhan commented on src/include/ASTree.h:
```diff
@@ -0,0 +1,104 @@
+
+struct Expression : public Statement { };
+
+class ExpressionGroup : public Expression {
```
I think this should all be `struct`s rather than `class`. I know they *technically* have a constructor method but they're basically just "dumb bags of data" (plus the parent `Statement` is a `struct`).

bobheadxi replied:
thats a good point! wasn't sure if we would need functions on them, but you're right they'll probably just hold stuff

rwblickhan commented on src/include/ASTree.h:
```diff
@@ -0,0 +1,104 @@
+class Function : public Statement {
+    std::string m_name;
+    std::vector<std::string> m_argNames;
+    std::unique_ptr<Statement> m_block;
```
How are block statements going to work? Personally I think I would prefer not having a separate block statement type and instead just representing a block statement as a vector of (pointers to) statements.

bobheadxi replied:
sorry the naming is a bit misleading, but `m_block` is just one `std::unique_ptr<Statement>`. I'm not sure if it should be a vector of statements or just a single statement though 🤔

rwblickhan replied:
Yeah, that's what I meant, I don't think we should have a block statement as a single `Statement` but rather as a vector of `Statement`s (ultimately I think we *will* need a vector, it's just a question of using it directly or having a `BlockStatement` struct).

bobheadxi replied:
a vector will probably do just fine instead of a `BlockStatement` - i'll turn this into a vector.

rwblickhan commented on src/include/ASTree.h:
```diff
@@ -0,0 +1,104 @@
+class Integer : public Literal {
+    int m_value;
+public:
+    explicit Integer(std::string value) : m_value(std::stoi(value)) { }
```
How will we handle errors here?

bobheadxi replied:
haven't thought too much about error handling, though we probably do want to at some point. I just assumed that anything the lexer says is an `Integer` would, in fact, be a valid integer:

```cpp

std::unique_ptr<Literal> Parser::parseLiteral() {
    std::unique_ptr<Literal> statement;
    switch (m_currentToken.type) {
        case TokenType::Integer:
            statement = std::make_unique<Integer>(
                    Integer(m_currentToken.literal));
            break;
        case TokenType::Float:
            statement = std::make_unique<Float>(
                    Float(m_currentToken.literal));
            break;
        case TokenType::String:
            statement = std::make_unique<String>(
                    String(m_currentToken.literal));
            break;
        case TokenType::True:
        case TokenType::False:
            statement = std::make_unique<Boolean>(
                    Boolean(m_currentToken.literal));
            break;
        default:
            // todo : exceptions, details
            printf("your token is shit");
            return std::unique_ptr<Literal>(nullptr);
    }
    getNextToken();
    return std::move(statement);
}

```

rwblickhan commented on src/include/ASTree.h:
```diff
@@ -0,0 +1,104 @@
+
+class FunctionCall : public Expression {
+    std::string m_functionName;
+    std::vector<std::unique_ptr<ExpressionGroup>> m_args;
```
Why a vector of expression groups in a function call? Shouldn't it be a vector of expressions and/or identifiers?

bobheadxi replied:
hm you're right. I'm not actually sure how to handle IDENT yet - will look into is soon, but this will probably be `Expression`

rwblickhan commented on src/include/Parser.h:
```diff
@@ -0,0 +1,44 @@
+    void buildTree();
+private:
+    Token getNextToken() {
+        return m_currentToken = m_lexer.getToken();
```
What is this method for? I imagine whoever owns the `Parser` won't need individual tokens from the `Lexer`, and anything inside `Parser` can just call `m_lexer.getToken()` directly.

(Also, I personally find assign-to-instance-variable-*and*-return confusing.)

bobheadxi replied:
i'm... actually not sure. one of the guides you sent said this was a buffer to allow lookahead, but that doesn't make sense to me anymore lol, I'll double check later

rwblickhan replied:
Oh yeah haha I didn't read those very closely. For lookahead I was thinking we could just read a statement at a time (e.g. until `EOL`?) and then any lookahead/error reporting/etc. would just be done for that statement (because in general statements shouldn't cross newlines). But there's some weirdness about for instance `if` statements and block statements. Probably something worth talking about tomorrow.

rwblickhan replied:
In particular I have a feeling @CoffeeTableEspresso has some thoughts on how to do that 😛

rwblickhan commented on src/include/ASTree.h:
```diff
@@ -0,0 +1,104 @@
+class String : public Literal {
+    std::string m_value;
+public:
+    explicit String(std::string value) : m_value(std::move(value)) { }
```
It's probably good practice to put a `const` qualifier on the `value` parameter. This would also mean you don't need to call `std::move` since the compiler should automatically reuse the same (immutable) string.

(Actually, it's probably a good idea to use `const` qualifiers wherever possible.)

bobheadxi replied:
will do!

rwblickhan commented:
Also we should probably rebase and force push to get rid of the `Lexer` conflicts. I can do that a bit later if you'd like.

bobheadxi committed:
Flesh out expression parsing

bobheadxi commented:
@rwblickhan rebased.

bobheadxi committed:
Address feedback

Remove header include in CMake
Turn ASTree classes into structs
Remove incorrect explicit declaration

bobheadxi committed:
Add util namespace for C++14-less make_unique

bobheadxi committed:
More parser work

rwblickhan committed:
Fix compile errors
